haskell-completions.el

```diff
@@ -220,36 +274,71 @@ result only if prefix length is not less than MINLEN."
+`haskell-completions-grab-identifier-prefix'."
+  (cl-destructuring-bind (beg end _pfx typ) prefix
+    (let ((candidates
+           (cl-case typ
+             ('haskell-completions-pragma-name-prefix
+              haskell-completions--pragma-names)
+             ('haskell-completions-ghc-option-prefix
+              haskell-ghc-supported-options)
+             ('haskell-completions-language-extension-prefix
+              haskell-ghc-supported-extensions)
+             (otherwise
+              haskell-completions--keywords))))
+      (list beg end candidates))))
+
+
+(defun haskell-completions-completion-at-point ()
+  "Provide a list of completion candidates.
+This function is supposed to be used in non-interactive context.
+It provides completions for haskell keywords, language pragmas,
+GHC's options, and language extensions, but not identifiers."
+  (let ((prefix (haskell-completions-grab-prefix)))
+    (haskell-completions--simple-completions prefix)))
 
-(defun haskell-completions-sync-completions-at-point ()
+(defun haskell-completions-sync-repl-completion-at-point ()
   "A `completion-at-point' function using the current haskell process.
 Returns nil if no completions available."
   (let ((prefix-data (haskell-completions-grab-prefix)))
     (when prefix-data
       (cl-destructuring-bind (beg end pfx typ) prefix-data
-        (let ((imp (eql typ 'haskell-completions-module-name-prefix))
-              lst)
-          (setq lst
-                (cl-case typ
-                  ;; non-interactive completions first
-                  ('haskell-completions-pragma-name-prefix
-                   haskell-completions-pragma-names)
-                  ('haskell-completions-ghc-option-prefix
-                   haskell-ghc-supported-options)
-                  ('haskell-completions-language-extension-prefix
-                   haskell-ghc-supported-extensions)
-                  (otherwise
-                   (when (and
-                          (not (eql typ 'haskell-completions-general-prefix))
-                          (haskell-session-maybe)
-                          (not
-                           (haskell-process-cmd (haskell-interactive-process))))
-                     ;; if REPL is available and not busy try to query it
-                     ;; for completions list in case of module name or
-                     ;; identifier prefixes
-                     (haskell-completions-sync-complete-repl pfx imp)))))
-          (when lst
-            (list beg end lst)))))))
+        (when (not (eql typ 'haskell-completions-general-prefix))
+          ;; do not complete things in comments
+          (if (cl-member
+               typ
+               '(haskell-completions-pragma-name-prefix
+                 haskell-completions-ghc-option-prefix
+                 haskell-completions-language-extension-prefix))
+              ;; provide simple completions
+              (haskell-completions--simple-completions prefix-data)
```

Comment:
As I understand there will be two completions hooks installed at the same time, one by `haskell-mode` other by `interactive-haskell-mode`. The non-interactive one will return completions for PRAGMA, GHC_OPTION and LANGUAGE, so there is no need to return those completions here again.

Unless I'm mistaken how the mechanism works.